pytorch_lightning/utilities/data.py

```diff
@@ -336,23 +352,28 @@ def wrapper(obj: DataLoader, *args: Any, **kwargs: Any) -> None:
             if param.name != "self" and param.kind not in (param.VAR_POSITIONAL, param.VAR_KEYWORD)
         ]
 
-        cls = type(obj)
+        set_arg_names = []
         for arg_name, arg_value in chain(zip(param_names, args), kwargs.items()):
-            if hasattr(cls, arg_name) and getattr(cls, arg_name).fset is None:
-                # the class defines a read-only (no setter) property of this name. it's likely that the implementation
-                # will set `self._arg_name = arg_value` in `__init__` which is the attribute returned by the `arg_name`
-                # property so we are fine skipping in that case
-                continue
-            setattr(obj, arg_name, arg_value)
+            if not hasattr(obj, "__" + arg_name):
+                # Set the value privately if it has not been set yet. This achieves two things:
+                # 1. The argument is the one that was passed in the outermost call
+                #    and not overriden by a call to `super().__init__()`
+                # 2. The argument is the passed value and does not get overwritten in `__init__()`
+                setattr(obj, "__" + arg_name, arg_value)
+                set_arg_names.append(arg_name)
+        if not hasattr(obj, "_set_arg_names"):
```

Comment:
Why this if?

And if we want it, shouldn't it also avoid the `for` loop entirely?

Reply:
So, ideally we want to only capture variables in the original, first call + `dataset`. The reason why we want `dataset` variable is just this line https://github.com/PyTorchLightning/pytorch-lightning/pull/12981/files#diff-f21f92490e6c1417367a48324385cf66c908d35fb4321a4b1b20e5ba410830f4R251

The same reason actually applies to changing the function `_get_all_subclasses`